src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs

```diff
@@ -251,10 +251,10 @@ private void OnTimer()
                 lock (s_counterGroupLock)
                 {
                     _timeStampSinceCollectionStarted = now;
-                    do
+                    for (int i = 0; i < 1000 && _nextPollingTimeStamp < now; i++)
```

Comment:
Where does this `1000` come from?

Also, why does this require a loop?  It can't just be done with some match to figure out the next time stamp?

Reply:
The 1000 is arbitrary with the assumption that `now` is not more than 1000 intervals in the future. If I recall correctly, this code is meant to prevent drift in the polling interval due to work time. For example, if the polling interval is 1 second (the smallest unit we allow), but the counter takes more than 1 second to collect, we don't want that work time to drift the next polling timestamp. The alternative would be to measure the time it took to calculate the counter then do a calculation comparing the work time, the interval, and `now` to get to the same result.

Reply:
Thanks.  I'm wondering why the loop is needed.  Why can't it be something along the lines of...
```C#
TimeSpan diff = now - _nextPollingTimeStamp;
if (diff > TimeSpan.Zero)
{
    _nextPollingTimeStamp += TimeSpan.FromMilliseconds(_pollingIntervalInMilliseconds * Math.Ceiling(diff.TotalMilliseconds / _pollingIntervalInMilliseconds));
}
```
Am I misreading what the loop is doing?

Reply:
I think you're right. That looks to achieve the same original intent. I prefer that over the loop. @noahfalk thoughts on using this change in terms of servicing? It's not quite as targeted, but I prefer _removing_ the loop over fixing the loop.

Reply:
The loop was added in https://github.com/dotnet/coreclr/pull/25864. It was an attempt to account for a situation where something (probably threadpool starvation) forces the timer to run significantly later than it should have. I.e. skip forward to the next counter if we run later than expected. https://github.com/dotnet/coreclr/pull/25864#discussion_r307058733

Reply:
I'm also fine removing the loop and replacing it with Stephen's calculation that computes the same result. Certainly whatever final form we use needs to handle _pollingIntervalInMilliseconds = 0. Currently for the loop we get infinite loop and for Stephen's math we'd get division by zero error. But as long as we've done something to handle that case gracefully then all good : )

Reply:
Re-reading my comment, if it seems like I was pushing for one outcome or another I wasn't. I was just providing context, I think the loopless method is a good solution.